drivers/ethernet/phy/phy_microchip_t1s.c

```diff
@@ -0,0 +1,561 @@
+{
+	const struct mc_t1s_config *cfg = dev->config;
+	int ret;
+
+	ret = mdio_read_c45(cfg->mdio, cfg->phy_addr, devad, reg, val);
+	/* @retval -ENOSYS if read using Clause 45 direct access is not supported */
+	if (ret == -ENOSYS) {
+		/* Read C45 registers using C22 indirect access registers */
+		ret = mdio_setup_c45_indirect_access(dev, devad, reg);
+		if (ret) {
+			return ret;
+		}
+
+		return mdio_read(cfg->mdio, cfg->phy_addr, MII_MMD_AADR, val);
+	}
+
+	return ret;
+}
+
+static int phy_mc_t1s_c45_write(const struct device *dev, uint8_t devad, uint16_t reg, uint16_t val)
+{
+	const struct mc_t1s_config *cfg = dev->config;
+	int ret;
+
+	ret = mdio_write_c45(cfg->mdio, cfg->phy_addr, devad, reg, val);
+	/* @retval -ENOSYS if write using Clause 45 direct access is not supported */
+	if (ret == -ENOSYS) {
+		/* Write C45 registers using C22 indirect access registers */
+		ret = mdio_setup_c45_indirect_access(dev, devad, reg);
+		if (ret) {
+			return ret;
+		}
+
+		return mdio_write(cfg->mdio, cfg->phy_addr, MII_MMD_AADR, val);
+	}
+
+	return ret;
+}
+
+static int phy_mc_t1s_get_link(const struct device *dev, struct phy_link_state *state)
+{
+	const struct mc_t1s_config *cfg = dev->config;
+	struct mc_t1s_data *data = dev->data;
+	struct phy_link_state old_state = data->state;
+	uint32_t value = 0;
+	int ret;
+
+	ret = phy_mc_t1s_read(dev, MII_BMSR, &value);
+	if (ret) {
+		LOG_ERR("Failed MII_BMSR register read: %d\n", ret);
+		return ret;
+	}
+
+	state->is_up = value & MII_BMSR_LINK_STATUS;
+	state->speed = LINK_HALF_10BASE_T;
+
+	if (memcmp(&old_state, state, sizeof(struct phy_link_state)) != 0) {
+		if (state->is_up) {
+			LOG_INF("PHY (%d) Link speed 10 Mb, half duplex\n", cfg->phy_addr);
+		}
+	}
+
+	return 0;
+}
+
+static int phy_mc_t1s_link_cb_set(const struct device *dev, phy_callback_t cb, void *user_data)
+{
+	struct mc_t1s_data *data = dev->data;
+
+	data->cb = cb;
+	data->cb_data = user_data;
+
+	if (data->cb) {
+		data->cb(dev, &data->state, data->cb_data);
+	}
+
+	return 0;
+}
+
+static void phy_monitor_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct mc_t1s_data *const data = CONTAINER_OF(dwork, struct mc_t1s_data, phy_monitor_work);
+	const struct device *dev = data->dev;
+
+	if (!data->cb) {
+		return;
+	}
+
+	phy_mc_t1s_get_link(dev, &data->state);
+
+	data->cb(dev, &data->state, data->cb_data);
+
+	/* Submit delayed work */
+	k_work_reschedule(&data->phy_monitor_work, K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+}
+
+/* Pulled from AN1760 describing 'indirect read'
+ *
+ * write_register(0x4, 0x00D8, addr)
+ * write_register(0x4, 0x00DA, 0x2)
+ * return (int8)(read_register(0x4, 0x00D9))
+ *
+ * 0x4 refers to memory map selector 4, which maps to MDIO_MMD_VENDOR_SPECIFIC2
+ */
+static int lan865x_indirect_read(const struct device *dev, uint16_t addr)
+{
+	uint16_t val;
```

Comment:
It is already implemented:

https://github.com/lmajewski/zephyr/blob/f0f91a8c8686c7cf761ca08e51e4b34d0a0f7a88/drivers/ethernet/eth_lan865x.c#L168